### PR TITLE
Add `DiscreteHilbert.states_to_local_indices` as it seems that it's used in multiple places

### PR DESCRIPTION
**Pull request by PhilipVinc**
This function convert every local degree of freedom which might be a whatever number to an integer index from 0 to ... Max.

This function already exists inside `nk.models.autoregressive` because it's needed there, and it cannot easily be extended for fermonic stuff. 
Moreover, this function came up as something that would be useful to implement #1209 concisely, so... here is it.

### CODE REVIEW

File: test/hilbert/test_hilbert.py [hunk 1]
```diff
@@ -325,6 +326,28 @@ def test_hilbert_index_discrete(hi: DiscreteHilbert):
...
 
+@partial(jax.jit, static_argnums=0)
+def _states_to_local_indices_jit(hilb, x):
+    return hilb.states_to_local_indices(x)
+
+
+@pytest.mark.parametrize("hi", discrete_hilbert_params)
+def test_states_to_local_indices(hi):
+    if isinstance(hi, nkx.hilbert.SpinOrbitalFermions):
+        pytest.xfail()
```
Comment (PhilipVinc) [inline test/hilbert/test_hilbert.py:337]: I'll leave it to @jwnys to implement this for SpinOrbitalFermions.

**Review by wdphy16 — APPROVED** [review]
Yes I always wanted to to this

**Review by gcarleo — APPROVED** [review]

**Comment by jwnys** [comment]
@PhilipVinc , why do you say "it cannot easily be extended for fermonic stuff."? Fermions are created by TensorHilbert of Focks space, both of which you have a "states_to_local_indices" implemented for..

File: test/hilbert/test_hilbert.py (hunk 1 above)
Comment (jwnys) [reply test/hilbert/test_hilbert.py:337]: @PhilipVinc , why do you say "it cannot easily be extended for fermonic stuff."? Fermions are created by TensorHilbert of Fock space, both of which you have a "states_to_local_indices" implemented for..

**Comment by PhilipVinc** [comment]
sorry, faulty keyboard + autocorrect transformed a `can->cann->cannot`. Of course it can.

**Comment by jwnys** [comment]
Ok, so i'll add this one-liner then.

**Pull request merged**